### DISCUSSION TITLE
v0.11.5

### DISCUSSION THREAD

**Discussion by fcakyon** [obss/sahi#757] [Announcements]
## What's Changed
* fixes a bug that prevents cuda device selection by @fcakyon in https://github.com/obss/sahi/pull/756

**Full Changelog**: https://github.com/obss/sahi/compare/0.11.4...0.11.5

<hr /><em>This discussion was created from the release <a href='https://github.com/obss/sahi/releases/tag/0.11.5'>v0.11.5</a>.</em>

**Comment by tureckova** [comment]
Hi,
I also run into an issue with device selection with mmdet model.
I am calling the function predict from my script and so far I did not set up the model_device parameter and everything works. But now I must set the model_device parameter to "cpu" or "cuda".  If I skip the model_device parameter a seg fault occurs during epoch loading:
```
load checkpoint from local path: epoch_24.pth
Segmentation fault (core dumped)
```
I have the following environment packages versions:
sahi==11.5, mmcv==1.7.0,  mmdet==2.26.0, torch==1.13.0+cu117.

The model_device parameter might be obligatory or at least print some warning if not used, since the seg fault is hard to track down and I spent quite some time discovering the cause.

**Reply by fcakyon** [reply]
@tureckova thanks for reporting. Can you tell me what is your pipeline using sahi so that I can recreate your issue?

Are you using `predict` function from `sahi.predict` or `get_sliced_prediction` function from `sahi.predict`?

**Reply by fcakyon** [reply]
@tureckova when you dont provide any device, sahi successfully selects `cuda:0`. You don't have to set it as `cpu` or `cuda`. I have just tested it with sahi==0.11.5, mmcv==1.7.0, mmdet==2.26.0, torch==1.12.1+cu113.

Maybe mmdet doesn't support torch 1.13 and cuda 11.7. Have you tried it with lower pytorch and cuda versions? (sahi is tested for cuda 11.3 and pytorch 1.12)

**Comment by tureckova** [comment]
Well, it is weird. It might be connected to cuda version since I encountered the problem after the upgrade.
If I activate **debug mode**, it works:
![image](https://user-images.githubusercontent.com/32820891/204516661-ad448038-7f6c-4a65-bbf2-aa51a21c130b.png)

But if I just **run** the script seg fault occurs. 
![image](https://user-images.githubusercontent.com/32820891/204517113-3808744c-b178-453b-bbbc-8251e4a6871b.png)

I tracked down your code (viz the printscreens), and it correctly gives device cuda:0, but in mmdet inference.py on line model.to(device) it is killed. I don't have time to dig deeper, but mmdet default train and test scripts works well for me. So I thought the bug is in sahi. I am pretty confused about where the bug came from.

**Reply by tureckova** [reply]
Ou, and it works on CPU well (but slowly).

**Reply by fcakyon** [reply]
The reason might be related to your pytorch and cuda versions

**Reply by yzfzzz** [reply]
This doesn't seem like a version issue, because my torch=1.11, python=3.8, mmcv=1.6.0, it also goes wrong

**Reply by yzfzzz** [reply]
when cu=116, do you have a test?

**Reply by fcakyon** [reply]
mmcv most probably dont support cu11.6. we currently only test for cu11.3, in near future: will be starting to test in cu11.4.

**Comment by yzfzzz** [comment]
Even though I use the same version as you, this problem still occurs and I think it may be related to cuda
![q](https://yzfzzz.oss-cn-shenzhen.aliyuncs.com/image/q.jpg)

![image-20221203202652509](https://yzfzzz.oss-cn-shenzhen.aliyuncs.com/image/image-20221203202652509.png)

This is my test code
```python
from sahi.predict import predict
from sahi import AutoDetectionModel

import faulthandler


faulthandler.enable()

# get batch predict result
result = predict(
    model_type='mmdet', # one of 'yolov5', 'mmdet', 'detectron2'
    model_path= '../mmdetection/tools/work_dirs/cascade_rcnn_val_r50_fpn_120x_coco_panda/best_bbox_mAP_epoch_115.pth', # path to model weight file
    model_config_path='../mmdetection/configs/panda/cascade_rcnn_r50_fpn_1x_coco_round1_panda.py', # for detectron2 and mmdet models
    model_confidence_threshold=0.5,
    model_device='cuda:0', # or 'cuda:0'
    source="./IMG_14_01.jpg", # image or folder path
    no_standard_prediction=True,
    no_sliced_prediction=False,
    slice_height=3000,
    slice_width=3000,
    overlap_height_ratio=0.5,
    overlap_width_ratio=0.5,
    export_pickle=False,
    export_crop=False,
)

```

**Reply by fcakyon** [reply]
What I mean is, mmcv do not support cuda 11.6 or 11.7. You have to use older cudatoolkit version for gpu imference.

**Reply by tureckova** [reply]
Well, the Cuda versions are certainly connected with the error. But mmdetection runs without trouble even with cuda 11.7, both training and test.
Interestingly, if I enable debug mode the seg fault does not occur and the prediction finishes without trouble. :astonished:

**Reply by fcakyon** [reply]
Then it might be related with your ide 🤔 I am not having any issues debugging with cuda 11.3 on vs code ide.

**Reply by kimlia545** [reply]
I had the same problem.
`
torch
'1.9.0+cu111'
mmdet
'2.25.1'
mmcv
'1.6.0'
`
In `get_sliced_prediction` device is "cuda" work, but in `predict`, when using "cuda", "Segmentation fault (core dumped)" error occurs and it works with "cpu".

**Reply by fcakyon** [reply]
What is your sahi version?

**Reply by kimlia545** [reply]
@fcakyon 
`sahi '0.11.13'`

**Reply by fcakyon** [reply]
Please try with latest sahi
